### PR TITLE
Zstd streaming and Hadoop Codec

### LINKED ISSUE #112: Streaming compression/decompression for Zstd
Hi. Is there support for Zstd streaming compression/decompression in this library (like in native zstd's streaming_compression/streaming_decompression examples)?

If not, can you mark this as an enhancement request please?

### PR DESCRIPTION
**Pull request by dain**
Resolve #112

### CODE REVIEW

**Comment by bondolo** [comment]
This appears to resolve #112

**Comment by hlship** [comment]
I'd love to see this merged and released; we're holding off on evaluating this library until the streaming API is available.

File: src/main/java/io/airlift/compress/zstd/XxHash64.java [hunk 1]
```diff
@@ -25,7 +32,150 @@
...
+            throws IOException
+    {
+        return hash(DEFAULT_SEED, in);
+    }
+
+    public static long hash(long seed, InputStream in)
+            throws IOException
+    {
+        XxHash64 hash = new XxHash64(seed);
+        byte[] buffer = new byte[8192];
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/XxHash64.java:169]: Why 8k?

File: src/main/java/io/airlift/compress/zstd/CompressionParameters.java [hunk 2]
```diff
@@ -237,9 +237,12 @@ public Strategy getStrategy()
...
     }
 
-    public static CompressionParameters compute(int compressionLevel, int inputSize)
+    public static CompressionParameters compute(int compressionLevel, int estimatedInputSize)
     {
-        CompressionParameters defaultParameters = getDefaultParameters(compressionLevel, inputSize);
+        CompressionParameters defaultParameters = getDefaultParameters(compressionLevel, estimatedInputSize);
+        if (estimatedInputSize < 0) {
+            return defaultParameters;
+        }
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/CompressionParameters.java:260]: This doesn't belong to this commit.

File: src/main/java/io/airlift/compress/zstd/CompressionParameters.java [hunk 3]
```diff
@@ -283,7 +286,7 @@ private static CompressionParameters getDefaultParameters(int compressionLevel,
     {
         int table = 0;
 
-        if (estimatedInputSize != 0) {
+        if (estimatedInputSize >= 0) {
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/CompressionParameters.java:304]: This doesn't belong to this commit.

File: src/main/java/io/airlift/compress/zstd/ZstdFrameDecompressor.java [hunk 4]
```diff
@@ -208,6 +217,342 @@ public int decompress(
...
+        final State state;
+        final boolean atStoppingPoint;
+        final FrameHeader frameHeader;
+        final int blockHeader;
+
+        final int inputConsumed;
+        final int outputBufferUsed;
+
+        final int inputRequired;
+        final int requestedOutputSize;
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/ZstdFrameDecompressor.java:242]: private

File: src/main/java/io/airlift/compress/zstd/ZstdFrameDecompressor.java [hunk 5]
```diff
@@ -208,6 +217,342 @@ public int decompress(
...
+
+    // current window buffer
+    private Object windowBase = new byte[0];
+    private long windowAddress = ARRAY_BYTE_BASE_OFFSET;
+    private long windowLimit = ARRAY_BYTE_BASE_OFFSET;
+    private long windowPosition = ARRAY_BYTE_BASE_OFFSET;
+
+    private XxHash64 partialHash;
+
+    public DecompressState partialDecompress(
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/ZstdFrameDecompressor.java:321]: Does this implementation share anything with the non-partial decompressor? If not, I'd move it into a separate class (e.g., ZstdIncrementalFrameDecompressor or ZstdResumableFrameDecompressor

File: src/test/java/io/airlift/compress/zstd/ZstdStreamDecompressor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.ByteBuffer;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ZstdStreamDecompressor
```
Comment (martint) [inline src/test/java/io/airlift/compress/zstd/ZstdStreamDecompressor.java:27]: What's the purpose of this class? Isn't it doing the same as the regular decompressor?

**Review by dain — COMMENTED** [review]
I rewrote the decompressor into a standalone, fully encapsulated, class, `ZstdIncrementalFrameDecompressor`

File: src/main/java/io/airlift/compress/zstd/ZstdIncrementalFrameDecompressor.java [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            final int outputLimit)
+    {
+        if (inputRequired > inputLimit - inputAddress) {
+            throw new IllegalArgumentException(format(
+                    "Required %s input bytes, but only %s input bytes were supplied",
+                    inputRequired,
+                    inputLimit - inputAddress));
+        }
+        if (requestedOutputSize > 0 && outputOffset >= outputLimit) {
+            throw new IllegalArgumentException("Output is ready, but output buffer is full");
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/ZstdIncrementalFrameDecompressor.java:114]: It'd be more accurate to say "Not enough space in output buffer to output <requestedOutputSize>"

**Review by martint — COMMENTED** [review]
These commits look good:

* Add streaming XxHash64
* Rename parameter to be more descriptive
* Support unknown input size in ZSTD CompressionParameters
* Add ZstdInputStream
* Add window size and block size to CompressionParameters

File: src/main/java/io/airlift/compress/zstd/ZstdOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                    blockSize,
+                    compressed,
+                    ARRAY_BYTE_BASE_OFFSET,
+                    compressed.length,
+                    context,
+                    lastChunk && blockSize == chunkSize);
+            outputStream.write(compressed, 0, compressedSize);
+            uncompressedOffset += blockSize;
+            chunkSize -= blockSize;
+        } while (chunkSize > 0);
```
Comment (martint) [inline src/main/java/io/airlift/compress/zstd/ZstdOutputStream.java:195]: Formatting

**Review by martint — APPROVED** [review]
